### PR TITLE
HDDS-8212. ozone admin printTopology command should have json output

### PR DESCRIPTION
**Pull request by Tejaskriya**
## What changes were proposed in this pull request?

Add the option for the command `ozone admin printTopology` to output as JSON

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8212

## How was this patch tested?

Added new Robot test. Also verified the JSON output using a docker-compose cluster:
```
bash-4.2$ ozone admin printTopology --json | more
[ {
  "ipAddress" : "172.25.0.104",
  "hostName" : "ozonesecure-ha-datanode3-1.ozonesecure-ha_ozone_net",
  "nodeState" : "HEALTHY",
  "operationalState" : "IN_SERVICE",
  "networkLocation" : "/default-rack",
  "ports" : {
    "HTTP" : 9882,
    "CLIENT_RPC" : 9864,
    "REPLICATION" : 9886,
    "RATIS" : 9858,
    "RATIS_ADMIN" : 9857,
    "RATIS_SERVER" : 9856,
    "RATIS_DATASTREAM" : 9855,
    "STANDALONE" : 9859
  }
}, {
--More--

bash-4.2$ ozone admin printTopology -o --json | more
[ {
  "ipAddress" : "172.25.0.104",
  "hostName" : "ozonesecure-ha-datanode3-1.ozonesecure-ha_ozone_net",
  "nodeState" : "HEALTHY",
  "operationalState" : "IN_SERVICE",
  "networkLocation" : "/default-rack"
}, {
  "ipAddress" : "172.25.0.103",
--More--

bash-4.2$ ozone admin printTopology -f --json | more
[ {
  "ipAddress" : "172.25.0.104",
  "hostName" : "ozonesecure-ha-datanode3-1.ozonesecure-ha_ozone_net",
  "nodeState" : "HEALTHY",
  "operationalState" : "IN_SERVICE",
  "networkLocation" : "/default-rack",
  "ports" : {
    "HTTP" : 9882,
    "CLIENT_RPC" : 9864,
    "REPLICATION" : 9886,
    "RATIS" : 9858,
    "RATIS_ADMIN" : 9857,
    "RATIS_SERVER" : 9856,
    "RATIS_DATASTREAM" : 9855,
    "STANDALONE" : 9859
  },
  "uuid" : "d590bb47-252e-4afc-8e40-b2fb27d62425"
}, {
--More--
```

### CODE REVIEW

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 1]
```diff
@@ -180,16 +204,119 @@ private String getAdditionNodeOutput(HddsProtos.Node node) {
...
-      System.out.println("    "
-          + node.getNodeOperationalStates(0) + "    " +
-          (node.getNodeID().getNetworkLocation() != null ?
-              node.getNodeID().getNetworkLocation() : "NA"));
-    });
+  private void printNodesWithLocation(Collection<HddsProtos.Node> nodes,
+                                      String state) throws IOException {
+    if (json) {
+      if (fullInfo) {
+        ArrayList<NodeTopologyFull> nodesJson = new ArrayList<>();
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:211]:
```suggestion
        List<NodeTopologyFull> nodesJson = new ArrayList<>();
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 2]
```diff
@@ -180,16 +204,119 @@ private String getAdditionNodeOutput(HddsProtos.Node node) {
...
+        nodes.forEach(node -> {
+          NodeTopologyFull nodeJson =
+              new NodeTopologyFull(
+                  DatanodeDetails.getFromProtoBuf(node.getNodeID()), state);
+          nodesJson.add(nodeJson);
+        });
+        System.out.println(
+            JsonUtils.toJsonStringWithDefaultPrettyPrinter(nodesJson));
+      } else {
+        ArrayList<NodeTopologyDefault> nodesJson = new ArrayList<>();
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:221]:
```suggestion
        List<NodeTopologyDefault> nodesJson = new ArrayList<>();
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 3]
```diff
@@ -180,16 +204,119 @@ private String getAdditionNodeOutput(HddsProtos.Node node) {
...
-          node.getNodeID().getHostName() + ")" +
-          ":" + formatPortOutput(node.getNodeID().getPortsList()));
-      System.out.println("    "
-          + node.getNodeOperationalStates(0) + "    " +
-          (node.getNodeID().getNetworkLocation() != null ?
-              node.getNodeID().getNetworkLocation() : "NA"));
-    });
+  private void printNodesWithLocation(Collection<HddsProtos.Node> nodes,
+                                      String state) throws IOException {
+    if (json) {
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:209]: Please merge multiple if else.

if (json && fullInfo) {.... return;}
else if(json && !fullInfo {... return;}

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 4]
```diff
@@ -153,13 +162,28 @@ private void printOrderedByLocation(List<HddsProtos.Node> nodes) {
...
+      locations.forEach(location -> {
+        tree.get(location).forEach(n -> {
+          NodeTopologyOrder nodeJson = new NodeTopologyOrder(n, nState,
+              state.get(n).toString());
+          nodesJson.add(nodeJson);
+        });
+      });
+      System.out.println(
+          JsonUtils.toJsonStringWithDefaultPrettyPrinter(nodesJson));
+    } else {
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:176]: Remove else, return from if statement.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 5]
```diff
@@ -136,7 +144,8 @@ public Class<?> getParentType() {
   // Format
   // Location: rack1
   //  ipAddress(hostName) OperationalState
-  private void printOrderedByLocation(List<HddsProtos.Node> nodes) {
+  private void printOrderedByLocation(List<HddsProtos.Node> nodes,
+                                      String nState) throws IOException {
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:148]:
```suggestion
                                      String state) throws IOException {
```

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 6]
```diff
@@ -180,16 +204,119 @@ private String getAdditionNodeOutput(HddsProtos.Node node) {
...
+    private String nodeState;
+    private String operationalState;
+    private String networkLocation;
+
+    NodeTopologyOrder(DatanodeDetails node, String nState, String opState) {
+      ipAddress = node.getIpAddress();
+      hostName = node.getHostName();
+      nodeState = nState;
+      operationalState = opState;
+      networkLocation = node.getNetworkLocation();
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:272]: Keep consistent as without json output
```
networkLocation = (node.getNodeID().getNetworkLocation() != null ?
    node.getNodeID().getNetworkLocation() : "NA")
```

**Review by ashishkumar50 — COMMENTED** [review]
@Tejaskriya Thanks for working on this, Please find few comments.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java (hunk 3 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java]: Thank you for the suggestion, I have incorporated it in my next commit.

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java (hunk 5 above)
Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java]: The variable name 'state' was already there, so I have renamed it to nodeStateStr. Is that okay?

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java]: I have fixed it now

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 7]
```diff
@@ -180,7 +204,32 @@ private String getAdditionNodeOutput(HddsProtos.Node node) {
 
   // Format "ipAddress(hostName):PortName1=PortValue1    OperationalState
   //     networkLocation
-  private void printNodesWithLocation(Collection<HddsProtos.Node> nodes) {
+  private void printNodesWithLocation(Collection<HddsProtos.Node> nodes,
+                                      String state) throws IOException {
+    if (json && fullInfo) {
```
Comment (ashishkumar50) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:209]: Please change like below:
```
if (json) {
   if (fullInfo) {
       .........
       return;
   }
.......... // without fullInfo handling
return;
}
```

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java]: I have changed it to this structure now

**Comment by ashishkumar50** [comment]
Thanks @Tejaskriya for updating patch. LGTM +1

**Review by ashishkumar50 — APPROVED** [review]

File: hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java [hunk 8]
```diff
@@ -136,10 +144,11 @@ public Class<?> getParentType() {
   // Format
   // Location: rack1
   //  ipAddress(hostName) OperationalState
-  private void printOrderedByLocation(List<HddsProtos.Node> nodes) {
+  private void printOrderedByLocation(List<HddsProtos.Node> nodes,
+                                      String state) throws IOException {
     HashMap<String, TreeSet<DatanodeDetails>> tree =
```
Comment (hemantk-12) [inline hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java:149]: nit: interface over implementation.
```suggestion
    Map<String, TreeSet<DatanodeDetails>> tree =  new HashMap<>();
    Map<DatanodeDetails, HddsProtos.NodeOperationalState> operationalState =
        new HashMap<>();
```

Comment (Tejaskriya) [reply hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/TopologySubcommand.java]: Thank you for the suggestions! I have made the changes.

File: hadoop-ozone/dist/src/main/smoketest/admincli/printTopology.robot [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+*** Test Cases ***
+Run printTopology
+    ${output} =         Execute          ozone admin printTopology
+                        Should Match Regexp     ${output}     State =
+
+Run printTopology as JSON
+    ${output} =         Execute             ozone admin printTopology --json
+    ${keys} =           Execute             echo '${output}' | jq -r '.[0] | keys'
+                        Should Contain      ${output}           ipAddress
```
[the remark this answers is not in the document]
Comment (Tejaskriya) [reply hadoop-ozone/dist/src/main/smoketest/admincli/printTopology.robot]: It was missing, I have added it now

**Review by hemantk-12 — COMMENTED** [review]
LGTM

**Review by aryangupta1998 — APPROVED** [review]
LGTM +1, Thanks @Tejaskriya for the PR, and @ashishkumar50, @hemantk-12 for the reviews!

**Pull request merged**